### DISCUSSION TITLE
Interchain Academy (Cosmos)

### DISCUSSION THREAD

**Discussion by shawntabrizi** [Polkadot-Blockchain-Academy/pba-content#598] [General]
I am thinking about signing up and take the Interchain Academy, which seems to be an online course for learning about developing on Cosmos.

https://academy.interchain.io

> The Interchain Developer Academy is a self-paced, seven-week training program covering the core aspects of the Interchain Stack, including the Cosmos SDK, CosmJS, and IBC.

**Comment by shawntabrizi** [comment]
There is a screening test, done by https://www.hackerrank.com

> To finish your application for the Interchain Developer Academy, you need to take a screening test. To help you maximize your chance of passing, at the end of this email, I’ve included some preparation tips and some example tests you can try out.
> 
> The test will evaluate your general knowledge of software development concepts and problem-solving skills. There are no blockchain-related challenges.
> 
> On average, it takes developers 82 minutes to finish the test. The maximum time given, which is 120 minutes, should be more than enough for you to complete it. However, keep in mind that you only have one chance to try and should answer as many questions as possible. Therefore, ensure that you are comfortable and in a place without any distractions.
> TAKE THE SCREENING TEST
> 
> Important dates: 
> 
>     The screening test closes on May 8th.
> 
>     We’ll let you know if you’ve passed the screening test by May 15th.
> 
>     The Academy program starts on May 18th.
> 
> 
> Good luck!
> 
> Syed and Interchain Developer Academy Team
> 
> Take a practice test 
> 
> The Academy's screening test is designed to assess your abilities as a web2 coder and we prioritize candidates who score highly, so it's a good idea to give yourself the best chance you can. Trying your hand at a few exercises might be useful preparation.
> 
> We recommend HackerRank's algorithm challenges – extensive examples of the kind of tasks you may be tested on. They include challenges with different difficulty levels (from easy to expert). Focusing on the “strings” subdomain should prove particularly useful.
> 
> In addition, HackerRank also provides data structure challenges. Consider paying special attention to challenges in the “arrays” subdomain.
> 
> Test format
> 
>     The screening test contains two types of questions: multiple choice and coding.
> 
>     You can attempt the questions in any order you like.
> 
>     Multiple choice questions ask you to select one or more correct answers.
> 
>     Questions are weighted differently. Two completed coding questions are worth as much as all the multiple choice questions combined.
> 
>     If you attempt a coding question but do not quite finish it, you may still get a partial score.
> 
>     Do not close the browser window/tab before you submit your final answers! If you do, we cannot guarantee that your work will be saved.
> 
> 
> More on the coding questions:
> 
>     Coding questions allow you to choose a language from a list. Pick the language you are most confident using.
> 
>     You are free to use different languages for different questions.
> 
>     If the code area contains the function signature, just complete the function alone (the platform takes care of the main function, headers, etc).
> 
>     If you are expected to write the full code, it is mentioned in the code area.
> 
>     All inputs are from STDIN and output to STDOUT. If you are using Java, use the classname "Solution".
> 
>     You can print to console to debug your code using the appropriate print command for each language (eg console.log for Javascript, cout for C++, printf for C, etc).
> 
>     You may prepare your preferred IDE with a simple function and empty unit tests so they are ready to go with your work. This will increase your response speed.
> 
> 
> Duration
> 
> A final reminder: you will have up to 120 minutes to complete the actual screening test and you only have one attempt to answer as many questions as possible, so make sure you are comfortable and in a distraction-free environment. The average time to complete the test is 82 minutes, so you should have plenty of time!
> 
> 
> Hopefully, now you can hit the ground running when you take the test!

**Reply by shawntabrizi** [reply]
I just finished the screening, and there was nothing related to blockchain.

Some questions about VMs, debugging, figuring out the output to some code (multi-choice), and them some algorithm coding questions.

Code they gave us was in various languages, we could code in any language. Not sure what they expect the average person going in the class to be able to do based on how general this exam was.

**Reply by nuke-web3** [reply]
Thanks for documenting this process here! Very curious what other possibly "competitive" programs to PBA offer.

ICYMI https://github.com/Polkadot-Blockchain-Academy/Rust-Entrance-Exam is the our private WIP monorepo to derive our
- Public Qualifier exam https://github.com/Polkadot-Blockchain-Academy/pba-qualifier-exam - a subset of (almost) completely autograded work as a minimal barrier to be considered in the application flow.
- "Assignment 0"  - a private set of resources with harder and more open ended tasks for those passing the qualifier around the interview stage of the application. Explicitly as an ungraded, highly encouraged, learning resource.

(Application flow for UCB is roughly: application form -> pass qualifier exam -> interviews -> offer extended -> placement locked)

**Comment by KiChjang** [comment]
Just based on the duration and format of the course, it would seem to me that theirs is much less intensive than PBA. I'm curious just how people learn effectively online, as we've seen engineers with many years of experience still struggling in person at previous PBAs. How do they assess students' progress?

I'm also interested to know just how they introduce IBC, since they don't have anything similar to XCM, but rather create message format standards for each and every use case, such as ICS-20 for fungible tokens and ICS-721 for NFTs. I'm sure that there is an interface where we can just purely send a blob of data to another chain (perhaps that's the packet API?), and with that, we could also try sending XCMs using IBC.

**Comment by shawntabrizi** [comment]
Just providing updates of the progress and layout of this academy:

I think their whole academy might be publicly accessible here: https://ida.interchain.io

> Hey Shawn Tabrizi!
> 
> Congratulations! You’ve passed the screening test and have been awarded a place in the Interchain Developer Academy.
> The program starts on May 18th when you will get full access to the program content and support in Discord.  
> Please keep a close eye on your inbox, before the start date you will receive two emails: 
> 1.	How to access the learning content
> 2.	How to access the private Academy space on the Cosmos Developer Discord
> 
> The rest of this email has important information so please read it carefully.
> 1.	Program kick-off is at 4 pm UTC on Thursday, May 18th.
> 2.	You'll receive an email to join the private Academy Discord channel.
> 3.	Expect to spend roughly 70 hours learning, and 20 hours on the final exam. We suggest you dedicate about 10 hours a week.
> 4.	There are no mandatory sessions.
> 5.	There is a quiz and a practical exercise during the first two weeks of the program. Neither is mandatory though we highly recommend both, they have proved to be an excellent indicator of overall success in previous cohorts.
> If you want to get a head start, you can already access the Academy platform. It features "Week 0" introductory content including an introduction to blockchain basics, tutorials on Golang and Docker, and an overview of some useful dev terms.
> 
> IMPORTANT: If you can no longer join the Academy, please reply to this email and let me know so I can reassign your seat to someone on the waiting list.

**Reply by shawntabrizi** [reply]
They have a ton of really nice follow up links throughout their material so far (section 0)

**Reply by shawntabrizi** [reply]
Really lots of really good follow up links to everything, and embedded videos to dive deeper. Feels like users really could spend a long time learning deeply about topics they touch only high level on.

**Comment by shawntabrizi** [comment]
First day of classes today:

> It’s time to start your Interchain journey! Here is the link to get going: https://ida.interchain.io/
>
> 
> But before you dive in, ask yourself: what kind of a learner are you?
> 
> * Do you like to learn the theory first, then build on that foundation?
> If that sounds like you, then we recommend you begin with the Week 1 content and continue through the program chronologically.
> 
> * Do you like to get your hands dirty right away, then work back to fill in the conceptual gaps?
> If this is your modus operandi, then we suggest you start with the Week 2 material now, then go back to read through Week 1 before you continue.
> 
> ⁉️ Whichever way you decide to approach the program, we highly recommend that you complete both Week 1 and Week 2!
> 
> Check out the following messages for an overview of both weeks 🙂
> 
> 🍀 And good luck! 🍀
> [Interchain Developer Academy](https://ida.interchain.io/)
> If you want to get to grips with the concepts behind Interchain both past and present, then where better to start than right here in Week 1?
> 
> https://ida.interchain.io/ida-course/LPs/week-1/
> 
> INTRODUCTION TO THE INTERCHAIN - A UNIVERSE OF LIMITLESS POSSIBILITIES
> 
> ▫️ Understand how Cosmos (aka “the Interchain”) fits in the overall development of blockchain technology.
> ▫️ Discover the Interchain Ecosystem.
> ▫️ Learn how to get your first ATOM and how to stake it.
> ▫️ Get an introduction to CometBFT and consensus in the Cosmos SDK.
> ▫️ Develop an understanding of the different components of the Cosmos SDK.
> ▫️ Read an introduction to relaying with the Inter-Blockchain Communication Protocol (IBC).
> 
> Don’t forget the Week 1 Quiz!
> 
> 🎯 At the end of the Week 1 content, we have prepared a quiz which you should submit before you move on to the next week. We highly recommend submitting the quiz as it is an opportunity to practice and demonstrate your engagement with the program. 🎯
> [Week 1 - A Universe of Limitless Possibilities | Interchain Develop...](https://ida.interchain.io/ida-course/LPs/week-1/)
> If you want to dive straight in and try building a blockchain-based project from scratch, then why not start your journey with the Week 2 content?
> 
> https://ida.interchain.io/ida-course/LPs/week-2/
> 
> TAKE YOUR FIRST STEPS INTO A WHOLE NEW COSMOS
> 
> ▫️ Set up your work environment.
> ▫️ Learn how to run a node and interact with a Cosmos SDK chain.
> ▫️ Work with Ignite CLI.
> ▫️ Start implementing a blockchain for a checkers game using Ignite CLI.
> 
> Don’t forget the Week 2 Exercise!
> 
> 🎯 At the end of the Week 2 content, we have prepared a voluntary exercise to help you check your progress. We recommend giving it a try – if you do, you should submit it before going back to check out Week 1. 🎯
> 
> Have fun taking your first steps toward building your own chain! 🚀
> [Week 2 - Get Started With Ignite Cli and Then Run Your Own Chain | ...](https://ida.interchain.io/ida-course/LPs/week-2/)

**Reply by shawntabrizi** [reply]
github repo for the academy is here: https://github.com/cosmos/sdk-tutorials/tree/master

**Reply by shawntabrizi** [reply]
Ignite CLI here: https://ignite.com/

**Comment by shawntabrizi** [comment]
It looks like their academy is built and run by a third party: https://b9lab.com/

These are the folks who seem to be coordinating everything, creating a gitlab for project submissions, etc...

**Reply by shawntabrizi** [reply]
They use Discord for their chat engine, and it has features like a Q/A forum built into the chat:

<img width="1727" alt="image" src="https://github.com/Polkadot-Blockchain-Academy/pba-content/assets/1860335/b1f7ccad-afdf-4a42-8d88-1a382b68f6e8">

**Reply by shawntabrizi** [reply]
lots of questions going through their Q&A.

It is quite easy to go through and find questions and follow up with conversations.

![image](https://github.com/Polkadot-Blockchain-Academy/pba-content/assets/1860335/c7871ea4-60bb-4de7-8c54-128689fc5e47)

**Reply by nuke-web3** [reply]
Do you think the UX of discord forums is better than stack exchange or discourse? Very curious what aspects of discord you like in comparison to what we use for pba (notion, element) and also the substrate/polkadot community.

As discourse/SE forum platforms are much more feature rich overall, especially for large public facing communities, I assume discord isn't a good replacement for them (for sure not self hostable or exportable on discord I presume)

**Reply by shawntabrizi** [reply]
I don't think that the Discord Q/A section is a replacement or alternative to Discourse or StackExchange.

Discord Q/A section is nice because it only includes the people involved in the academy, and lives right next to the general chat which everyone uses throughout the academy to coordinate.

Also, these questions are contextual to the academy, versus people asking questions which would be specific to the academy, in public forums where not everyone would have context of the question.

This is much better than the shared google doc or just asking questions into the general chat which get lost to scroll.

**Reply by shawntabrizi** [reply]
The b9 labs guys are not super technical, yet do a reasonable job at leading the mechanics of the academy.

![image](https://github.com/Polkadot-Blockchain-Academy/pba-content/assets/1860335/479694e1-0721-4e00-adf5-f0347abcc5ea)

ive seen them:

- make daily posts to the discord
-  answer some questions, more answers from some of the more technical guys
- send out emails / reminders
- fix up documentation / courses caught by students
- make videos which are walkthroughs of their tutorials

**Comment by kianenigma** [comment]
I would love to hear your feedback about participating in this in Lisbon.

As for things like discord, once PBA is a separate entity, it would be free to switch to a new messaging platform.

**Comment by shawntabrizi** [comment]
They have released their final exam today. Here is the readme for that project:

---

# Project description

You should read this whole page before attacking the problem, including the part about Docker and your working copy.

This is the beginning of a larger project dedicated to the future payments of [toll roads](https://en.wikipedia.org/wiki/Toll_road). The project, built with Ignite 0.22.1 and CosmJS 0.28.13, is far from complete.

However it already contains:

1. A single `SystemInfo`. It is meant to keep track of system-wide information, in this case, the ID of the next road operator created. It was created with:

    ```sh
    $ ignite scaffold single SystemInfo nextOperatorId:uint --no-message
    ```

2. A map of `RoadOperator`s. It was created with:

    ```sh
    $ ignite scaffold map RoadOperator name token active:bool
    ```

    The default behavior of the scaffolding command is to have the index of the operator come from `MsgCreateRoadOperator`. However, `index` was removed from `MsgCreateRoadOperator`. That's because when a user creates a new road operator, the user does not choose the ID. Instead, it is chosen by the system on creation. After rebuild, various compilation errors were "fixed" in a lazy way.

3. The third-party Protobuf files necessary to compile to Typescript. And the Protobuf compiler executable. Some incopmlete CosmJS Typescript files.

## Do now

Before doing any work, and before you forget, do these steps:

1. Clone the repository.
2. Build the Docker image with:

    ```sh
    $ docker build -f Dockerfile-exam . -t exam_i
    ```

3. Read this document in full.

This is not to save you time. Instead this is to save you mistakes and surprises down the road.

## Your work

The following steps are in order of increasing difficulty:

1. Adjustments on system info.
2. Adjustments on road operators.
3. Add a new store type: User vaults.
4. Compile the Protobuf files to Typescript and complete the missing parts in the incomplete CosmJS files.

The tests have been divided into different packages to avoid compilation errors while your project is incomplete.

### On system info

Adjust `x/tollroad/types/genesis.go` so that the `x/tollroad/genesis_test.go` tests pass. To confirm, run:

```sh
$ go test github.com/b9lab/toll-road/x/tollroad
```

### On road operators

When a road operator is created, its ID has to be taken from `SystemInfo`. For this part, you are going to work only in `x/tollroad/keeper/msg_server_road_operator.go` and in it, only adjust the `CreateRoadOperator` function body. **Not the function signature, not another function, not another file.**

This is what you have to implement:

1. Make sure that the new road operator has its ID taken from `SystemInfo`.
2. Have this ID returned by the message server function.
3. Make sure the next id in `SystemInfo` is incremented.
4. Emit an event with the expected type and attributes.

To confirm, run:

```sh
$ go test github.com/b9lab/toll-road/x/tollroad/roadoperatorstudent
```

Look into the `x/tollroad/roadoperatorstudent/msg_server_road_operator_test.go` file to see what is expected, in particular the details of the expected event.

### On user vaults

This part requires more work.

With the operators in place, it is time to move your attention to the users of the road operators. Users are going to keep some tokens in escrow with the operators. The idea is that road operators will eventually be allowed to transfer to themselves some of the tokens that the users have put into escrow "with them".

To keep track of **which user** has put **how much** of **which token denomination** in escrow with **which operator**, you have to add a new type named `UserVault`.

The user vault object has exactly 4 fields, not less, not more. In Protobuf it should be:

```protobuf
message UserVault {
    string owner = 1; 
    string roadOperatorIndex = 2; 
    string token = 3; 
    uint64 balance = 4; 
}
```

The user vault object's key in the map is the combination of, and in this order, `owner, roadOperatorIndex, token`. This means, for instance, that the future keeper function to get a user vault has this signature:

```go
func (k Keeper) GetUserVault(ctx sdk.Context, owner string, roadOperatorIndex string, token string) (val types.UserVault, found bool)
```

In effect, `balance` is the only field that is not part of the object's key.

Additionally, the message to create this vault object should not have the `owner` field, as it is in effect picked from the `creator` field. In Protobuf, the create message is exactly:

```protobuf
message MsgCreateUserVault {
    string creator = 1;
    string roadOperatorIndex = 2;
    string token = 3;
    uint64 balance = 4;
}
```

It should not be allowed to create another user vault with the same key of `owner, roadOperatorIndex, token`.

Similarly, the message to update the vault picks the owner in the `creator` field. In Protobuf, it looks like `MsgCreateUserVault`:

```protobuf
message MsgUpdateUserVault {
    string creator = 1;
    string roadOperatorIndex = 2;
    string token = 3;
    uint64 balance = 4;
}
```

`balance` is the only field that can be updated.

And again, the message to delete a vault:

```protobuf
message MsgDeleteUserVault {
    string creator = 1;
    string roadOperatorIndex = 2;
    string token = 3;
}
```

With these objectives in mind, your tasks are as follows.

#### Scaffold the type

So your first task is to add a new mapped type named `UserVault` with `ignite scaffold map`. If you do it right, you can:

1. Use a single Ignite command.
2. Do minor adjustments on Protobuf objects to match as per the above.
3. Fix compilation errors that appear after successive rebuilds.
4. Adjust the functions in `x/tollroad/client/cli/tx_user_vault.go` as per the change in how things need to be called, i.e. replacing `owner` with creator.

#### Handle tokens

With the data structure, the keeper and the messages stuff done, your second task is to handle the tokens by calling the bank.

We have prepared mocks in `testutil/mock_types/expected_keepers.go`. The `MockBankEscrowKeeper` is a mock of a yet-to-be-created `type BankEscrowKeeper interface`. You have to:

1. In `x/tollroad/types/expected_keepers.go`, declare this interface.
2. Have it declare the two bank functions that transfer tokens between module and accounts.

With this done, apply what you learned in the course so that:

1. Your keeper has the permissions and the capability to transfer tokens between users and its module.
2. You use the mock `MockBankEscrowKeeper` in the **test** keeper initialization. To find out where this takes place, follow through the set up of tests.
3. Your keeper has to be able to actually transfer tokens between the user and the module:
    1. On creating the vault:
        1. The `balance` amount is transferred from the user to the module.
        2. If the user does not have enough tokens, then it should return an error. See the tests for the details of messages.
        3. If the amount is `0` then it returns an error.
    2. On deleting the vault:
        1. The `balance` amount is transferred from the module to the user.
        1. If the module does not have enough tokens, it should panic. See the tests for the details of messages.
    3. On updating the vault:
        1. If the balance field in the message is higher than the current vault balance, the difference is transferred from the user to the module. And should return an error if it is not possible.
        2. If the balance field in the message is lower than the current vault balance, the difference is transferred from the module to the user. And should panic if it is not possible.
        3. If the balance field in the message is `0` then it returns an error, because conceptually, this should be a deletion.
4. If your keeper function receives an error when calling the bank, it should return this error unmodified, like so:

    ```go
    err = k.bank.SendCoins...
    if err != nil {
        return nil, err
    }
    ```

    This is just to make sure that you pass the tests.

#### Checking it all

To confirm, run:

```sh
$ go test github.com/b9lab/toll-road/x/tollroad/uservaultstudent
```

The tests are in two files:

1. `x/tollroad/uservaultstudent/msg_server_user_vault_test.go` that runs unit tests with mocks. The mocks confirm that the bank was called as expected.
2. `x/tollroad/uservaultstudent/tx_user_vault_test.go` that starts a full app. The tests also confirm that the toll-road module has the expected balance.

Check in these files to see the details of what is expected.

### On CosmJS

Start by compiling the Protobuf files into Typescript.

* The necessary third party files are already there. For instance [`proto/google/api/annotations.proto`](proto/google/api/annotations.proto).
* The [`protoc`](Dockerfile-exam#L41) executable is in your Docker image. It works if you are on an Intel 64 bits or Apple M1 platform. In particular, if you are using another platform, you will have to adjust `Dockerfile-exam` or download `protoc` for [your local environment](https://github.com/protocolbuffers/protobuf/releases/tag/v21.7). After you have built the image, confirm that it works with:

```sh
$ docker run --rm -it exam_i protoc
```

To confirm that you generated the correct Typescript files, confirm there are no compilation errors in [`queries.ts`](client/src/modules/tollroad/queries.ts).

With the objects created, you can move to the task of filling in the missing parts in the following files:

* [`client/src/types/tollroad/events.ts`](client/src/types/tollroad/events.ts) has functions that throw `"Not implemented"`.
* [`client/src/modules/tollroad/queries.ts`](client/src/modules/tollroad/queries.ts) has a function that throws `"Not implemented"`.
* [`client/src/types/tollroad/messages.ts`](client/src/types/tollroad/messages.ts) has missing message objects.
* [`client/src/tollroad_signingstargateclient.ts`](client/src/tollroad_signingstargateclient.ts) has functions that throw `"Not implemented"`.

At the core, the tests are regular NPM tests: [`client/test/integration/one-run.ts`](client/test/integration/one-run.ts). However, as you can see from the many `before` clauses, it calls the Ignite faucet to populate the test accounts. So you cannot just run `npm test` as usual.

You need to run [`testing-cosmjs.sh`](testing-cosmjs.sh):

```sh
$ ./testing-cosmjs.sh
```

It starts an `ignite chain serve` before `npm test`, and stops it afterwards.

## Preparing your working copy

This is your personal repository, it is named like `IDA-P3-final-exam/student-projects/YOUR_NAME-exam-code` and this is where you need to upload your work.

To prepare your working copy:

1. Clone your repository:

    ```sh 
    $ git clone https://YOUR_NAME@git.academy.b9lab.com/ida-p3-final-exam/student-projects/YOUR_NAME-exam-code.git
    ```

    Where you have replaced `YOUR_NAME` with its actual value in both places. If you are unsure of your username, you can find it [here](https://git.academy.b9lab.com/-/profile/account). For instance:

    ```sh
    $ git clone https://alice@git.academy.b9lab.com/ida-p3-final-exam/student-projects/alice-exam-code.git 
    ```

2. Work on the exercise. When you are happy with the result(s). Commit (the messages matter for you but not for the grading) and push.

    ```sh
    $ git add THE_FILES_YOU_ARE_COMMITTING
    $ git commit -m "Add my submission."
    $ git push
    ```

3. You can keep `push`ing as many further commits as you want before the deadline.

After the submission deadline, we will run the tests on your latest version of the `main` branch, not on an intermediate commit you made.

If you wish to use the SSH protocol instead of supplying your username and password over HTTPS to perform Git operations like clone/pull/push, you can learn how to [handle SSH keys](https://docs.gitlab.com/ee/user/ssh.html#generate-an-ssh-key-pair) and add your SSH keys [here](https://git.academy.b9lab.com/-/profile/keys).

## Looking at test files

Make sure you **read the section about Docker** before you jump into running anything.

For your convenience, here are the `go test` commands that the grading scripts will run:

```sh
$ go test github.com/b9lab/toll-road/x/tollroad
$ go test github.com/b9lab/toll-road/x/tollroad/roadoperatorstudent
$ go test github.com/b9lab/toll-road/x/tollroad/uservaultstudent
```

Each of them is actually launched from a script, respectively:

```sh
$ ./x/tollroad/testing.sh
$ ./x/tollroad/roadoperatorstudent/testing.sh
$ ./x/tollroad/uservaultstudent/testing.sh
```

Inside each of them you can see how much weight is given to each test.

The NPM tests are launched from `./testing-cosmjs.sh` and each of the 4 tests has the same weight.

In turn, these 4 `testing` scripts are launched from this script:

```sh
$ ./score.sh
```

When `score.sh` has run well, irrespective of the completeness of the exercise, it outputs a score such as:

```txt
FS_SCORE:4%
```

This is your score as we will record it.

Inside `score.sh` and the 4 `testing`, you can see how much weight is given to each individual test and to each part of the exercise. Namely:

* About `SystemInfo`: [1](score.sh#L12)
* About `RoadOperator`: [2](score.sh#L22)
* About `UserVault`: [4](score.sh#L32)
* About ComsJS: [2](score.sh#L48)

You can run these scripts (inside Docker or not, see below) as many times as you want, and see what score that would give you. We do not track how many times you run them and do not keep a score on our end when _you_ run them.

Passing all tests gives you 100%, passing none gives you 0%, and from the start you get 4%. Passing a single one gives you a score that depends on the weights applied.

How do you test the outcome? That's the object of testing it in Docker. The paragraphs also discuss read-only files.

## Running tests in Docker

This is the right way to run the tests, as it will let you test the outcome just the way we will test it after you have submitted your work to us. Install [Docker](https://docs.docker.com/engine/install/).

This project was actually built within a container described by the Docker file `Dockerfile-exam`. Do the following to run your tests in Docker too.

* Create the image:
  
    ```sh
    $ docker build -f Dockerfile-exam . -t exam_i
    ```

    It will take time because it downloads the [Go](Dockerfile-exam#L43) and [NPM](Dockerfile-exam#L47) dependencies onto the image so that the containers do not have to do it every time you run the tests.

    In fact you should create the image **as soon as you cloned it**. Indeed, it makes [a copy](Dockerfile-exam#L42) of the current directory into `/original` so as to have a snapshot of the files that have to remain read-only (more about that below).

* Run the scoring in a throwaway container just like you did earlier:

    ```sh
    $ docker run --rm -it -v $(pwd):/exam -w /exam exam_i ./score.sh
    ```

* Run the scoring like we will do when you have submitted your work:

    ```sh
    $ docker run --rm -it -v $(pwd):/exam -w /exam exam_i /original/score-ci.sh
    ```

Notes on `score-ci.sh`:

* It is launched from the `/original` folder which is within the container, and then runs `score.sh` on `/exam` which is your working folder.
* It starts by [overwriting](score-ci.sh#L5) all the files that should be [read-only](fileconfig.yml#L7), by taking the original ones in `/original`. So be sure to make a commit so as not to lose work. If it appears that you have edited a file that is supposedly read-only, then you have to make further commits to undo your changes.
* This is the script that we run on our Gitlab runner and it will overwrite any read-only file that you updated.
* The score you get here is the closest to the score we will get when running it on our side.

Now, if you want something more permanent to debug your code:

* Build a reusable container:

    ```sh
    $ docker create --name exam -i -v $(pwd):/exam -w /exam -p 1317:1317 -p 4500:4500 -p 26657:26657 exam_i
    $ docker start exam
    ```

* Test a single part on the reusable container. Connect to it:

    ```sh
    $ docker exec -it exam bash
    ```

    Then in this connected shell:

    ```sh
    $ go test github.com/b9lab/toll-road/x/tollroad
    ```

    Or, in the same shell:

    ```sh
    $ ./score.sh
    ```

    Or even:

    ```sh
    $ /original/score-ci.sh
    ```

It is possible that you could do without Docker if you have all the right tools and versions installed already. But better use Docker and `/original/score-ci.sh` at the end at least to avoid any surprises.

## Official grading

To grade your project, we run the same `score-ci.sh` file via a Gitlab pipeline. In fact, you can see yourself the status of the pipeline's jobs [here](../../jobs) or from the "CI/CD -> Jobs" menu in the navigation on the left.

1. If a job named `test` has a _Passed_ status, it means that the job itself ran as expected. To see your score, click on it and look at the bottom of the log where you can find the `FS_SCORE` information.
2. If a job named `test` has a _Failed_ status, it means that the job failed, usually because of a timeout, and there is no score to collect. You can send it back to the job queue by clicking the relaunch button. If it fails twice in a row, you should alert us.

Have a look inside the [`fileconfig.yml`](./fileconfig.yml) file. Under the `readonly_paths` label is the list of files and folders that our script will overwrite in your repository before running the tests. If you modify any of these files, you may have inconsistent results between what you see _on your machine_ and _on the pipeline's job_.

After the submission deadline, we will make sure all `test` jobs _Passed_ and will collect all your scores.

## Submission checklist

Before you consider your work done, check everything off this list:

1. You committed your work in the `main` branch.
2. You pushed your `main` branch to the `main` branch on our Academy Gitlab server.
3. On the [jobs](../../jobs) page, your latest job has completed and at the end of it, you see the same `FS_SCORE` that you got when running locally on your computer. If the [jobs](../../jobs) link does not work:

    * It is not because the page does not exist.
    * You need to open it from the repository itself.

## Cleanup

* If you have worked a bit on the project, are not happy and want to start over, you can use Git's:

    ```sh
    $ git stash -u && git stash drop
    ```

* If you want to remove the Docker elements you created:

    ```sh
    # If you created and started the reusable container
    $ docker stop exam
    # If you created the reusable container
    $ docker rm exam
    # If you created the image
    $ docker rmi exam_i
    ```

Please do not publish your solution to the outside world so that we can reuse the exercise or parts of it.

Good luck.

**Reply by shawntabrizi** [reply]
Note that the score is entirely calculated with a script that even the student can run. It seems there is no reason any student shouldn't get 100% if they can follow the instructions and ask enough questions.
